### PR TITLE
Game logic

### PR DESCRIPTION
**Pull request by balerum03**
- Suggested branch name game_logic.
- Add real logic for your game - get rid of all inline comments and mocked logic used in previous milestone.
- As main.rb file should be responsible for user interaction, this is the only(!) one place where puts/print and get.chomp can be used.
- Think how you can encapsulate diffrent parts of game logic in classes. Do not create one class for everything.
- Read about DRY rule and use it.
- Think about the diffrence between attr_readers/attr_writers and atrr_acessors. Use atrr_acessors only if you really need it.
- Remember about single responsibility rule. Always.
- Remember to cover edge cases: think what player could try to do to cheat the game and make sure that your program does not allow to perform invalid action.

### CODE REVIEW

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far, team. 👍 

- The game works properly. Congratulations on the correct game logic!

## However, there are some changes required 👀 🧰 🔨 🔧 

- In order for your `bin/main` file to be properly executable, the first line must be `#!/usr/bin/env ruby`, not `# Variable definition area`. Please fix this.
- It's time to organize your code into classes! Create one class for each part involved in the game logic (e.g. board, players, game, etc...), and then group the methods involved in inside their corresponding class.
- These classes should not be a part of the `bin/main` file. You should create new files inside the `lib` directory for these game logic resources.
- You should access these resources from the `bin/main` file using `require` or `require_relative`.
- Only the `bin/main` file should contain `puts` and `gets chomp` calls. The classes in `lib` should only handle game logic, not input/output functions!
- Make sure that you make proper use of `attr_readers/attr_writers/attr_accessors` in your classes.


### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

File: bin/main [hunk 1]
```diff
@@ -16,27 +17,25 @@ player2_name = gets.chomp
...
       puts "it's #{player1_name} turn"
       display_board(board)
-      board_input = gets.chomp
+      $board_input = gets.chomp
     else
       puts "it's #{player2_name} turn"
       display_board(board)
-      board_input = gets.chomp
+      $board_input = gets.chomp
     end
```
Comment (kubilaycaglayan) [inline bin/main]: Move logic into the lib folder. There shouldn't be any logic in main.rb.

File: bin/main [hunk 2]
```diff
@@ -1,73 +1,73 @@
...
-    swap_on_board(i, given_number)
-  	puts "good job \n\n"
-    if given_number.to_i == 7
-      puts "you have won!"
-      break
-    elsif given_number.to_i == 8
-      puts "it's a draw"
-      break
+while itr < 9
+  given_number = interface.input_request(itr, players.player1, players.player2, board.board)
```
Comment (kubilaycaglayan) [inline bin/main]: I understand why you use 9 here. It feels logical at first sight. But we prefer to use infinite loops with terminating conditions. Please, switch this into an infinite loop and form a proper control mechanism to end that loop.

File: bin/main [hunk 3]
```diff
@@ -1,73 +1,73 @@
...
-    if i.even?
-      $board[given_number.to_i - 1] = 'X'
-    elsif i.odd?
-      $board[given_number.to_i - 1] = 'O'
+  def input_request(num, player1_name, player2_name, board)
+    if num.even?
+      puts "it's #{player1_name} turn"
+    else
+      puts "it's #{player2_name} turn"
     end
```
Comment (kubilaycaglayan) [inline bin/main]: Let's move the logic into the lib folder. We don't want any logic inside the main.rb.

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    elsif board[3] == board[4] && board[3] == board[5]
+      true
+    elsif board[6] == board[7] && board[6] == board[8]
+      true
+    elsif board[0] == board[3] && board[0] == board[6]
+      true
+    elsif board[1] == board[4] && board[1] == board[7]
+      true
+    elsif board[2] == board[5] && board[2] == board[8]
+      true
```
Comment (kubilaycaglayan) [inline lib/board.rb]: Would you consider doing something like this?

```ruby
  def check_win(player_symbol)
    # for rows
    all_equal(board[0..2], player_symbol) ||
    all_equal(board[3..5], player_symbol) ||
    all_equal(board[6..8], player_symbol)

    # and analogical as above for columns and diagonals
  end

  def all_equal(subarray, player_symbol)
    return true if subarray.uniq.size == 1 && subarray.uniq == player_symbol
  end
```

File: bin/main [hunk 5]
```diff
@@ -1,73 +1,73 @@
...
-    swap_on_board(i, given_number)
+    board.swap_on_board(itr, given_number)
+  end
+  if board.winner?(board.board)
+    system('cls') || system('clear')
+    interface.display_board(board.board)
+    if itr.even?
+      puts "The winner is #{players.player1}"
+    else
+      puts "The winner is #{players.payer2}"
```
Comment (kubilaycaglayan) [inline bin/main]: Move the logic into the lib folder please. 
**And for all the other logic, do the same please.**

**Review by kubilaycaglayan — CHANGES_REQUESTED** [review]
Hello Team :wave: 
Status: Changes Required :recycle: 

Great job on making the changes! :star: 
And the gameplay is great too:star2: 

Let's cover the details together :arrow_down: 

- [ ] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

:bulb: I've added inline comments, _see the comments below please_.

- [ ] DRY rule should be respected.

**DRY means don't repeat yourself and it is a great implementation principle.**
:bulb: I've added inline comments for your winner check method _check comments below please_

Thanks for your efforts and time :+1: 
With all my respect... :pray:

File: lib/board.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    board.count('X') + board.count('O') == 9
+  end
+
+  def winner?(itr)
+    sign = if itr.even?
+             'X'
+           else
+             'O'
+           end
+    win_arr.each do |arr|
```
Comment (expjazz) [inline lib/board.rb]: - [x] This should be @win_arr. I got an error on my console without it.

image

**Review by expjazz — CHANGES_REQUESTED** [review]
## Status: Changes Required :recycle: 

Great job implementing the changes requested by my colleagues :ok_hand: 

However, you have a bug. Kindly check the inline comments for more info.

### Please make the required changes and resubmit for a new code review once you are ready
Good Luck and Happy Coding :pray:

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Expedito Andrade).:smile:_

**Comment by krys2fa** [comment]
# Status: Project Approved! 🥇 

Congratulations on making all the requested changes and getting this project approved. 👍 
You can now merge this branch into the master branch and go on with your next project.

💡  Do not forget to submit your project completion form after merging this branch.

#### Nice-To-Haves:
The following are a few **optional** suggestions to make your game better:
- [ ] [OPTIONAL] Your game currently accepts numbers as player names. Checking for names will be great.
- [ ] [OPTIONAL] Your game ends as soon as there is a draw or a win. It will be better to have the game ask the players to play again or to end the game.

All the best team,
Chris

**Pull request merged**